app/anteplus/msg_ante.go

```diff
@@ -0,0 +1,119 @@
+package anteplus
+
+import (
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/cosmos-sdk/x/authz"
+	govkeeper "github.com/cosmos/cosmos-sdk/x/gov/keeper"
+	govtypes "github.com/cosmos/cosmos-sdk/x/gov/types"
+
+	claimtypes "github.com/crescent-network/crescent/v5/x/claim/types"
+)
+
+// initial deposit must be greater than or equal to 50% of the minimum deposit
+var minInitialDepositFraction = sdk.NewDecWithPrec(50, 2)
+
+type MsgFilterDecorator struct {
+	govKeeper *govkeeper.Keeper
+	cdc       codec.BinaryCodec
+}
+
+func NewMsgFilterDecorator(cdc codec.BinaryCodec, govKeeper *govkeeper.Keeper) MsgFilterDecorator {
+	return MsgFilterDecorator{
+		govKeeper: govKeeper,
+		cdc:       cdc,
+	}
+}
+
+func (d MsgFilterDecorator) AnteHandle(
+	ctx sdk.Context, tx sdk.Tx,
+	simulate bool, next sdk.AnteHandler,
+) (newCtx sdk.Context, err error) {
+	msgs := tx.GetMsgs()
+	if err = d.ValidateMsgs(ctx, msgs); err != nil {
+		return ctx, err
+	}
+
+	return next(ctx, tx, simulate)
+}
+
+func (d MsgFilterDecorator) ValidateMsgs(ctx sdk.Context, msgs []sdk.Msg) error {
+	var minInitialDeposit sdk.Coins
+	validateMsg := func(msg sdk.Msg, nested bool) error {
+		// mempool(check tx) level msg filter
+		if ctx.IsCheckTx() {
+			switch msg := msg.(type) {
+			// prevent messages with insufficient initial deposit amount
+			case *govtypes.MsgSubmitProposal:
+				if minInitialDeposit.Empty() {
+					fmt.Println("update")
+					depositParams := d.govKeeper.GetDepositParams(ctx)
+					minInitialDeposit = CalcMinInitialDeposit(depositParams.MinDeposit, minInitialDepositFraction)
+				}
+
+				if !msg.InitialDeposit.IsAllGTE(minInitialDeposit) {
+					return sdkerrors.Wrapf(sdkerrors.ErrInsufficientFunds, "insufficient initial deposit amount - required: %v", minInitialDeposit)
+				}
```

Comment:
I think it'd be better to have multiple ante handlers instead of having a single ante handler which does many different checks.

Reply:
I also think splitting multiple ante handlers would reduce code complexity and make it easier to read, but I thought it would be more efficient to combine it because it would result in duplicated processes for msgs iteration in tx and unpacking authz nested msg